services/code_hosting/git_hub_service.rb

```diff
@@ -1,131 +1,409 @@
+        return self.cache[[repo, branch].join("/")] if cache && self.cache && !self.cache[[repo, branch].join("/")].nil? && self.cache[[repo, branch].join("/")].kind_of?(Hash)
+        path = File.join(path, repo, branch)
+        begin
+          git_path = File.join(path, repo.split("/").last)
+          # This triggers the check of an existing repo in the given path,
+          # we recover from the error making the clone and checkout
+          Git.open(git_path)
+          fastfile_path = self.fastfile_path(root_path: git_path)
+          fastfile = Fastlane::FastfileParser.new(path: fastfile_path)
+          fastfile_config = {}
+          # TODO: FastfileParser provides a nice and clean tree of the Fastfile given,
+          # but there's a specific case when there might be lanes or actions out of
+          # a platform scope, in those cases the key is nil. We sanitize that so the user
+          # still can select lanes without platform. But this should be done from the
+          # FastfileParser side.
+          fastfile.tree.each_key do |key|
+            if key.nil?
+              fastfile_config[:no_platform] = fastfile.tree[key]
+            else
+              fastfile_config[key.to_sym] = fastfile.tree[key]
+            end
+          end
+          fastfile_json_path = File.join(path, "fastfile.json")
+          FileUtils.touch(fastfile_json_path) unless File.exist?(fastfile_json_path) && !File.zero?(fastfile_json_path)
+          File.write(fastfile_json_path, JSON.pretty_generate(fastfile_config))
+          self.cache[[repo, branch].join("/")] = fastfile, fastfile_config
+          return fastfile, fastfile_config
+        rescue ArgumentError
+          self.clone(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential,
+            path: path
+          )
+          fastfile, fastfile_config = self.peek_fastfile_configuration(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential,
+            path: path,
+            cache: cache
+          )
+          return fastfile, fastfile_config
+        rescue RuntimeError
+          # This is because no Fastfile config was found, so we cannot go further.
+          return {}
+        end
+      end
+
+      # Class method that shallow clones the repo on the target branch.
+      # @param [String] repo_url
+      # @param [String] branch
+      # @param [GithubProviderCredential] provider_credential
+      # @return [Git::Base]
+      def clone(repo_url: nil, name: nil, branch: nil, sha: nil, provider_credential: nil, path: nil)
+        repo = self.repo_from_url(repo_url)
+
+        folder_name = name || repo.split("/").last
+        clone_path = path || self.temp_path
 
-    GitHubService.status_context_prefix = "fastlane.ci: "
+        FileUtils.rm_rf(File.join(clone_path, folder_name)) if File.directory?(File.join(clone_path, folder_name))
```

Comment:
I suggest you ad intermediate temporary named variables so that we understand why `File.join(clone_path, folder_name)` is needed, so people can read the code and understand what is happening easier.

Reply:
That's a good point